### ISSUE TITLE
```go test -race``` detects race condition when using sonic. But using inbuilt `encoding/json` does not.

### ISSUE THREAD

**Issue by bigfoot31** [bytedance/sonic#45]
```WARNING: DATA RACE
Write at 0x00c0002c5870 by goroutine 20:
  github.com/bytedance/sonic/internal/caching.(*_ProgramMap).insert()
      /Users/probhonjon/Dev/go1.16/pkg/mod/github.com/bytedance/sonic@v0.0.0-20210628042511-8fe152d152e5/internal/caching/pcache.go:111 +0x136
  github.com/bytedance/sonic/internal/caching.(*_ProgramMap).add()
      /Users/probhonjon/Dev/go1.16/pkg/mod/github.com/bytedance/sonic@v0.0.0-20210628042511-8fe152d152e5/internal/caching/pcache.go:82 +0xc5
  github.com/bytedance/sonic/internal/caching.(*ProgramCache).Put()
      /Users/probhonjon/Dev/go1.16/pkg/mod/github.com/bytedance/sonic@v0.0.0-20210628042511-8fe152d152e5/internal/caching/pcache.go:142 +0xa9
  github.com/bytedance/sonic/decoder.findOrCompile()
      /Users/probhonjon/Dev/go1.16/pkg/mod/github.com/bytedance/sonic@v0.0.0-20210628042511-8fe152d152e5/decoder/pools.go:104 +0x431
  github.com/bytedance/sonic/decoder.decodeTypedPointer()
      /Users/probhonjon/Dev/go1.16/pkg/mod/github.com/bytedance/sonic@v0.0.0-20210628042511-8fe152d152e5/decoder/primitives.go:28 +0x44
  github.com/bytedance/sonic/decoder.(*Decoder).Decode()
      /Users/probhonjon/Dev/go1.16/pkg/mod/github.com/bytedance/sonic@v0.0.0-20210628042511-8fe152d152e5/decoder/decoder.go:63 +0x384
  github.com/bytedance/sonic.UnmarshalString()
      /Users/probhonjon/Dev/go1.16/pkg/mod/github.com/bytedance/sonic@v0.0.0-20210628042511-8fe152d152e5/sonic.go:41 +0x44f
  github.com/bytedance/sonic.Unmarshal()
      /Users/probhonjon/Dev/go1.16/pkg/mod/github.com/bytedance/sonic@v0.0.0-20210628042511-8fe152d152e5/sonic.go:36 +0x3d1
```

Logic: lots of goroutines call a `helper function`. This `helper function` calls sonic to extract json.

**Comment by chenzhuoyu** [comment]
This might be a bug within the RCU map implementation. We will investigate it later.

**Comment by AsterDY** [comment]
@bigfoot31  Would you please give a detail UT on this issue? Although we've been aware of where the problem is but can't reproduce the same result using old codes

**Comment by bigfoot31** [comment]
## To replicate:
1. run `go test -race`
2. replace sonic with encoding/json and then run `go test -race`

## Versions:
- `go version go1.16.5 darwin/amd64`

**file: main.go**
```go
package main

import (
	"fmt"

	"github.com/bytedance/sonic"
)

func main() {
	fmt.Println("plain")
}

func ExtractJson(idx int, body interface{}) {
	j := []byte(exampleJson[idx])

	// REPLACE sonic WITH json
	err := sonic.Unmarshal(j, &body)
	if err != nil {
		fmt.Println(err.Error())
		return
	}
}

var exampleJson1 = `{
"page_info": { "total_count": 1, "page_num": 1 },
"code": 0,
"data": [
	{
	"name": "dp-91b905d33f-5559868569-z8tvk",
	"node_ip": "10.245.11.66",
	"psm": "A.B.C",
	"containers": [
		{
			"status": "Running",
			"mem_usage_pct": 15.0,
			"name": "dp-91b905d33f",
			"cpu_usage_pct": 3.8
		},
		{
			"status": "Running",
			"mem_usage_pct": 13.0,
			"name": "dp-91b905d33f",
			"cpu_usage_pct": 2.1
		}
	]
	}
],
"error": ""
}`

var exampleJson2 = `{
	"page_info": { "total_count": 1, "page_num": 1 },
	"code": 0,
	"data": [
		{
		"name": "dp-91b905d33f-5559868569-z8tvk",
		"psm": "P.Q.R",
		"containers": [
			{
				"status": "Running",
				"mem_usage_pct": 15.0,
				"name": "dp-91b905d33f"
			},
			{
				"status": "Running",
				"name": "dp-91b905d33f",
				"cpu_usage_pct": 2.1
			}
		]
		}
	],
	"error": ""
  }`

var exampleJson = []string{exampleJson1, exampleJson2}
```

**file: main_test.go**
```go
package main

import (
	"fmt"
	"sync"
	"testing"
)

//
//
// model definition for pod information
type podInfoResp struct {
	Data    *[]inputPodInfo `json:"data"`
	Code    *int            `json:"code"`
	Message *string         `json:"message"`
	Error   *string         `json:"error"`
}
type inputPodInfo struct {
	PSM            *string          `json:"psm"`
	NodeName       *string          `json:"name"`
	NodeIP         *string          `json:"node_ip"`
	ContainersInfo []containersInfo `json:"containers"`
}
type containersInfo struct {
	Name        *string  `json:"name"`
	Status      *string  `json:"status"`
	CPUUsagePct *float64 `json:"cpu_usage_pct"`
	MemUsagePct *float64 `json:"mem_usage_pct"`
}

func TestExtracJson(t *testing.T) {

	wg := sync.WaitGroup{}

	resultChan := make(chan podInfoResp, 2)

	wg.Add(1)
	go func() {
		defer wg.Done()
		var model podInfoResp
		ExtractJson(0, &model)
		resultChan <- model
	}()

	wg.Add(1)
	go func() {
		defer wg.Done()
		var model podInfoResp
		ExtractJson(1, &model)
		resultChan <- model
	}()

	var results []podInfoResp
	for i := 0; i < 2; i++ {
		results = append(results, <-resultChan)
	}

	close(resultChan)

	fmt.Println(*(*results[0].Data)[0].PSM)
	fmt.Println(*(*results[1].Data)[0].PSM)
}
```

**Comment by chenzhuoyu** [comment]
Fixed in #49.
